File: src/main/scala/com/github/phisgr/gatling/generic/action/CodeActionBuilder.scala
package com.github.phisgr.gatling.generic.action

import com.github.phisgr.gatling.generic.check.{CodeCheck, ResponseExtract}
import io.gatling.commons.validation.Validation
import io.gatling.core.action.Action
import io.gatling.core.action.builder.ActionBuilder
import io.gatling.core.check.CheckBuilder
import io.gatling.core.session.{Expression, Session}
import io.gatling.core.structure.ScenarioContext

import scala.util.Try

case class CodeActionBuilder[T](
  requestName: Expression[String],
  f: (Session, Try[T] => Unit) => Unit,
  private val checks: List[CodeCheck[T]] = Nil,
  isSilent: Boolean = false
) extends ActionBuilder {
  override def build(ctx: ScenarioContext, next: Action): Action = {
    val resolvedCheck = if (checks.isEmpty) CodeCheck.defaultCheck[T] else checks
    new CodeAction[T](
      requestName = requestName,
      f = f,
      checks = resolvedCheck,
      isSilent = isSilent,
      statsEngine = ctx.coreComponents.statsEngine,
      clock = ctx.coreComponents.clock,
      next = next
    )
  }

  def check[X](
    extract: Try[T] => Validation[Option[X]])(
    checks: (CheckBuilder.Find[ResponseExtract, Try[T], X] => CodeCheck[T])*
  ): CodeActionBuilder[T] = {
    val checkBuilder = ResponseExtract.extract(extract)
    copy(checks = this.checks ::: checks.map(_.apply(checkBuilder)).toList)
  }

  def checkSeq[X](
    extract: Try[T] => Validation[Option[Seq[X]]])(
    checks: (CheckBuilder.MultipleFind[ResponseExtract, Try[T], X] => CodeCheck[T])*
  ): CodeActionBuilder[T] = {
    val checkBuilder = ResponseExtract.extractMultiple(extract)
    copy(checks = this.checks ::: checks.map(_.apply(checkBuilder)).toList)
  }

  def silent: CodeActionBuilder[T] = copy(isSilent = true)
}

File: gatling-kt-ext/src/main/kotlin/com/github/phisgr/gatling/kt/generic/CodeActionBuilder.kt
package com.github.phisgr.gatling.kt.generic

import com.github.phisgr.gatling.generic.Predef
import com.github.phisgr.gatling.generic.check.CodeCheck
import com.github.phisgr.gatling.generic.check.ResponseExtract
import com.github.phisgr.gatling.kt.internal.ActionCheckBuilder
import com.github.phisgr.gatling.kt.internal.codeCheck
import io.gatling.javaapi.core.CheckBuilder
import io.gatling.javaapi.core.internal.Converters
import scala.util.Try
import com.github.phisgr.gatling.generic.action.CodeActionBuilder as CodeActionBuilderS
import io.gatling.core.check.CheckBuilder as CheckBuilderS

class CodeActionBuilder<Res>(
    private val wrapped: CodeActionBuilderS<Res>,
    private val checks: List<CodeCheck<Res>> = emptyList(),
) : ActionCheckBuilder<
    CodeActionBuilder<Res>,
    Try<Res>,
    Try<Res>,
    CodeCheck<Res>
    >(codeCheck) {
    fun silent(): CodeActionBuilder<Res> = CodeActionBuilder(wrapped.silent(), checks)

    override fun asScala(): CodeActionBuilderS<Res> = wrapped
        .copy(
            wrapped.requestName(),
            wrapped.f(),
            Converters.toScalaSeq(checks).toList(),
            wrapped.isSilent
        )

    @Suppress("UNCHECKED_CAST")
    override fun buildCheck(builder: CheckBuilder): CodeCheck<Res> =
        Predef.checkBuilder2CodeCheck(
            builder.asScala() as CheckBuilderS<ResponseExtract, Try<Res>>
        )

    override fun addChecks(checks: List<CodeCheck<Res>>): CodeActionBuilder<Res> =
        CodeActionBuilder(wrapped, this.checks + checks)
}
